### PR TITLE
feat(terminal): close shell terminals automatically

### LINKED ISSUE #8975: Please support closing term window when command exits
- `nvim --version`:

```
NVIM v0.3.1
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-XJHVyS/neovim-0.3.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -DDISABLE_LOG -Wdate-time -D_FORTIFY_SOURCE=2 -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim-XJHVyS/neovim-0.3.1/build/config -I/build/neovim-XJHVyS/neovim-0.3.1/src -I/usr/include -I/build/neovim-XJHVyS/neovim-0.3.1/build/src/nvim/auto -I/build/neovim-XJHVyS/neovim-0.3.1/build/include
Compiled by team+vim@tracker.debian.org

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

- Vim (version: 8.1 + patches 1-320) behaves differently? Yes
- Operating system/version: Debian, latest unstable
- Terminal name/version: `GNOME Terminal 3.30.0 using VTE 0.54.0`
- `$TERM`: `xterm-256color`

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:term
exit
```

### Actual behaviour
nvim displays `[Process exited 0]` in the terminal and waits for a keypress.

### Expected behaviour
Terminal automatically closes when running a default shell.

vim's default feels right here: `:term` with no specified command will close the terminal buffer when the command exits, while `:term somecommand` will keep the terminal buffer open after the command exits (such as to allow reviewing output).

### LINKED ISSUE #14986: do not show [Process exited] in finished terminals
The `[Process exited]` prompt is extremely handy when running non-interactive commands like `make` or `make test`, as it lets you examine the output before the window disappears.

However, for running interactive applications like `lazygit` or `htop`, it is just an additional keystroke you need to press to close the window.

I have this keybinding set up to launch `lazygit` in a new tab:

```vim
" Open lazygit in a new tab
if has("nvim")
    nnoremap <Leader>g :tabnew<CR>:terminal lazygit<CR>i
endif
```

When I'm done I press `q`, which exits `lazygit`. But then I have to press another key to close the window. This isn't a huge deal of course, but it does irk me since it's a completely unnecessary keypress. I will never need to examine the terminal output after `lazygit` exits, so it would feel a lot nicer if I could simply press `q` and go right back to the file I was editing.

The same is true when I just want to open a quick terminal to examine something in the command line. When I press <kbd>Ctrl</kbd><kbd>d</kbd> to exit the terminal, it feels unnecessary to have to press another key to close the window.

I would love an option when opening a new terminal to not prompt for a keypress after the command exits.

Maybe:

```vim
:terminal <noprompt> lazygit
```

---

I tried hacking together a workaround using `TermClose`, but it seems that event fires before the prompt is displayed, so I'm not sure if it can work:

```
autocmd TermClose term://*lazygit <CR>
```

image

### PR DESCRIPTION
**Pull request by gpanders**
When starting a terminal buffer with no arguments (i.e. `:terminal`), automatically close the buffer if the shell exits without an error.

Closes #8975.

### CODE REVIEW

**Comment by gpanders** [comment]
I'm going to see if I can also find a way to close https://github.com/neovim/neovim/issues/5176 as well.

File: src/nvim/ex_docmd.c [hunk 1]
```diff
@@ -9504,6 +9504,13 @@ static void ex_terminal(exarg_T *eap)
   }
 
   do_cmdline_cmd(ex_cmd);
+
+  if (*eap->arg == NUL) {
+    do_cmdline_cmd("autocmd TermClose <buffer> "
+                   " if !v:event.status | "
+                   "   exec 'bdelete! ' .. expand('<abuf>') | "
+                   " endif");
```
Comment (justinmk) [inline src/nvim/ex_docmd.c:9512]: Probably should be global rather than `<buffer>` local. And we can declare it in a `nvim_` augroup so that users can opt-out.

Comment (gpanders) [reply src/nvim/ex_docmd.c:9512]: I don’t think it should be global as then it would apply to all `termopen` calls rather than just `:terminal`. The former provides finer grained control so we should make fewer assumptions and allow the user to do whatever they want, while `:terminal` is a convenience function that should have more convenient defaults.

My 2¢.

**Comment by gpanders** [comment]
> I'm going to see if I can also find a way to close #5176 as well.

https://github.com/neovim/neovim/pull/15442 takes care of #5176, so if we can find a way to make the `TermClose` autocommand somehow enter the `terminal_enter` code path we can kill two birds with one stone.

File: src/nvim/ex_docmd.c (hunk 1 above)
Comment (justinmk) [reply src/nvim/ex_docmd.c:9512]: Hmm. Thinking about how a user could opt-out of this in its current form: `autocmd TermClose` is done after the `do_cmdline_cmd(ex_cmd);` above. So this is setup after TermOpen... how would users opt-out?

> I don’t think it should be global as then it would apply to all `termopen` calls

Makes sense, but OTOH consistent behavior is easier to reason about, and to document. Clear documentation (one-liner that can be pasted), somewhat addresses the termopen() case, and avoids contortions for the :terminal case.

A global autocmd or augroup solves the user config case, but doesn't solve the termopen() case for scripts:  can't selectively opt-out, except by `:noautocmd call termopen(...)` ... maybe that's reasonable...

Comment (gpanders) [reply src/nvim/ex_docmd.c:9512]: > A global autocmd or augroup solves the user config case, but doesn't solve the termopen() case for scripts: can't selectively opt-out, except by `:noautocmd call termopen(...)` ... maybe that's reasonable...

I think that's reasonable, especially if it's documented.

If we do make it a global autocommand then we may want to resolve https://github.com/neovim/neovim/pull/15401 first so that we can use whatever default autocommand framework we decide on.

Comment (gpanders) [reply src/nvim/ex_docmd.c:9512]: I've been thinking about this more and two things occurred to me:

1) Making it a global autocommand means we can't use the heuristic of whether or not an argument was provided to `:terminal` to determine if we should automatically close the buffer or not. Right now, we create this autocommand if there is no argument and we are therefore using a default shell. But if, for instance, a user runs `:terminal ls` then we *do not* want to automatically close the buffer when the job completes. If we use a global autocommand we will need another flag or something to indicate which state we are in.

2) I wonder if it's not actually more common that users may want to opt-out of the auto-close behavior on a per-job/per-terminal instance, rather than opting out globally. This also provides plugins finer grained control to opt out of this behavior, which wouldn't be possible (or at least, more complex) with a global autocommand.

For these two reasons I think we ought to stick with a buffer-local autocommand. What do you think?

Comment (justinmk) [reply src/nvim/ex_docmd.c:9512]: > 1\. ... But if, for instance, a user runs `:terminal ls` then we _do not_ want to automatically close the buffer when the job completes. If we use a global autocommand we will need another flag or something to indicate which state we are in.

yeah, that's where I was going with [this comment](https://github.com/neovim/neovim/pull/15440#discussion_r693233441):

> In a global TermOpen autocmd, that could possibly be derived from the `term://` buffer name, or maybe we should expose the args in `v:event`. Or `nvim_get_chan_info(&channel)`

re (2):

> This is much more straightforward to do with a buffer-local autocommand. This would also provide plugins finer-grained control to opt out of this behavior,

Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`

Comment (gpanders) [reply src/nvim/ex_docmd.c:9512]: > Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`.

The plugin would need to disable the global autocommand in that case though, which the user may not want. The only way around this would be to add some kind of buffer-local variable that acts as a flag (e.g. `let b:terminal_autoclose = v:false`) that the plugin could set. But if we're doing that, why not just use a buffer-local autocommand?

I guess I disagree that a global autocommand is less messy, it paints with too broad of a brush so we end up needing to introduce further measures to account for edge cases and exceptions.

Comment (justinmk) [reply src/nvim/ex_docmd.c:9512]: >> Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`.

^ this was regarding buffer-local autocommand

Comment (gpanders) [reply src/nvim/ex_docmd.c:9512]: Ah cool, sorry for the misunderstanding 😅

**Comment by justinmk** [comment]
Different idea: what if we only set this on `TermEnter ` (terminal buffer version of "InsertEnter")? I'm trying this in my config:

```
autocmd TermOpen *
  \ autocmd TermEnter <buffer=abuf> ++once
  \ autocmd TermClose <buffer=abuf> if !v:event.status | exe 'bdelete '..expand('<abuf>') | endif
```

**Comment by gpanders** [comment]
> Different idea: what if we only set this on `TermEnter ` (terminal buffer version of "InsertEnter")? I'm trying this in my config:
> 
> ```
> autocmd TermOpen *
>   \ autocmd TermEnter <buffer=abuf> ++once
>   \ autocmd TermClose <buffer=abuf> if !v:event.status | exe 'bdelete '..expand('<abuf>') | endif
> ```
> 
> So far it works as desired with these cases:
> 
>     * `:te top` (interactive)
> 
>     * `:te ls` (non-interactive)
> 
>     * `:te` (default shell)

This won't work if the user uses `autocmd TermOpen * startinsert` (which the docs mention in `:h :terminal`). I'd wager this is not that uncommon.

**Comment by justinmk** [comment]
> This won't work if the user uses `autocmd TermOpen * startinsert` (which the docs mention in `:h :terminal`). I'd wager this is not that uncommon.

good catch. At this point, it seems like this is a requirement:

> .... expose the `:terminal` args in `v:event`. Or `nvim_get_chan_info(&channel)`

**Comment by gpanders** [comment]
> .... expose the :terminal args in v:event. Or nvim_get_chan_info(&channel)

Fortunately, `f_termopen()` already parses out the command argument; we just need to store that somewhere.

I'm thinking of adding a new `const char *cmd` field to `struct Terminal`. Do you have a better suggestion?

I've already looked at just using `proc->argv[0]`, but this memory is freed when the process exits, which can happen even if the terminal buffer is still alive (that wouldn't really be that helpful anyway, because in the event someone uses e.g. `:terminal ls`, `proc->argv[0]` would still just be the shell).

**Comment by justinmk** [comment]
> I'm thinking of adding a new `const char *cmd` field to `struct Terminal`. Do you have a better suggestion?

`nvim_get_chan_info()` could include an `argv` field. PR: https://github.com/neovim/neovim/pull/15537 . Only works for running jobs currently, because...

> I've already looked at just using `proc->argv[0]`, but this memory is freed when the process exits, which can happen even if the terminal buffer is still alive

Yeah good point. I wonder if we can just link job lifetime to the channel lifetime. Is there a reason we need to free jobs before channels? Probably not...

> **EDIT:** Wait wait, if we're not doing the global autocommand, isn't it just sufficient to check `*eap->arg == NUL`?

Ideally we avoid using internal magic, so that these decisions can be made by scripts.

**Comment by gpanders** [comment]
With #15537 the check simply becomes

```vim
if nvim_get_chan_info(&channel).argv ==# [&shell]
```

**Comment by gpanders** [comment]
Lint failure is on `screen.c` (why?), otherwise everything looks good!

**Comment by clason** [comment]
> Lint failure is on `screen.c` (why?), otherwise everything looks good.

Because dundargoc taught uncrustify new tricks, tested them on `screen.c`, and nobody told the linter about it yet ;)

TL;DR: unrelated

**Comment by zeertzjq** [comment]
Note: Because of #16257, this solution leaks memory with every terminal closed with `:bdelete` on `TermClose`.

**Review by justinmk — APPROVED** [review]
Very happy to see this! We can continue to improve little details, but this looks like a monotonic improvement meanwhile.

File: runtime/lua/vim/_editor.lua [hunk 2]
```diff
@@ -1107,13 +1107,26 @@ end
...
-  vim.api.nvim_create_autocmd({ 'cmdwinenter' }, {
+  vim.api.nvim_create_autocmd({ 'TermClose' }, {
+    group = nvim_terminal_augroup,
+    desc = 'Automatically close terminal buffers when started with no arguments and exiting without an error',
+    callback = function(args)
+      if vim.v.event.status == 0 then
+        local info = vim.api.nvim_get_chan_info(vim.bo[args.buf].channel)
+        local argv = info.argv or {}
+        if #argv == 1 and argv[1] == vim.o.shell then
+          vim.cmd({ cmd = 'bdelete', args = { args.buf }, bang = true })
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:1124]: This will close the current split window (if any). That's probably unwanted in many cases, so we need either of:

- open `:terminal` in split by default
- add a `:bdelete` variant that doesn't close the current split (or hack it here using `b#`) https://github.com/neovim/neovim/issues/2434

Comment (gpanders) [reply runtime/lua/vim/_editor.lua:1124]: I made https://github.com/neovim/neovim/pull/15442 a while ago for the first option. Even though it would probably break a lot of things, I think I’d still prefer the split-by-default behavior 😬

Comment (justinmk) [reply runtime/lua/vim/_editor.lua:1124]: Do you mean https://github.com/neovim/neovim/pull/15427 ? Unless it breaks fugitive I think we should try it.

Comment (gpanders) [reply runtime/lua/vim/_editor.lua:1124]: Oops, yes, thanks. Updated my original comment.
